set-matrix-zeroes/HC-kang.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * https://leetcode.com/problems/set-matrix-zeroes
+ * T.C. O(r * c)
+ * S.C. O(r + c)
+ */
+function setZeroes(matrix: number[][]): void {
+  const r = matrix.length;
+  const c = matrix[0].length;
+
+  const zeroRows = new Set<number>();
+  const zeroCols = new Set<number>();
+
+  for (let i = 0; i < r; i++) {
+    for (let j = 0; j < c; j++) {
+      if (matrix[i][j] === 0) {
+        zeroRows.add(i);
+        zeroCols.add(j);
+      }
+    }
+  }
+
+  for (let i = 0; i < r; i++) {
+    for (let j = 0; j < c; j++) {
+      if (zeroRows.has(i) || zeroCols.has(j)) {
+        matrix[i][j] = 0;
+      }
+    }
+  }
+}
+
+/**
+ * T.C. O(r * c)
+ * S.C. O(1)
+ */
+function setZeroes(matrix: number[][]): void {
+  const r = matrix.length;
+  const c = matrix[0].length;
+
+  let firstRowHasZero = false;
+  let firstColHasZero = false;
+
+  if (matrix[0].some((val) => val === 0)) {
+    firstRowHasZero = true;
+  }
+
+  if (matrix.some((row) => row[0] === 0)) {
+    firstColHasZero = true;
+  }
+
+  for (let i = 1; i < r; i++) {
+    for (let j = 1; j < c; j++) {
+      if (matrix[i][j] === 0) {
+        matrix[i][0] = 0;
+        matrix[0][j] = 0;
+      }
+    }
+  }
+
+  for (let i = 1; i < r; i++) {
+    if (matrix[i][0] === 0) {
+      matrix[i].fill(0);
+    }
+  }
+
+  for (let j = 1; j < c; j++) {
+    if (matrix[0][j] === 0) {
+      matrix.forEach((row) => (row[j] = 0));
+    }
+  }
```

Comment:
(질문) 반복문을 하나로 합쳐 진행해도 좋을 것 같은데 두 개로 나누신 이유가 따로 있으실까요?

Reply:
넵 질문주셔서 감사합니다!

`matrix`의 가로, 세로 사이즈가 다를 수 있어서 분리 하였는데, 취향에 따라 아래와 같이 표현할 수는 있을 것 같습니다.

```ts
for (let i = 1; i < Math.max(r, c); i++) {
  if (matrix[i]?.[0] === 0) matrix[i].fill(0);
  if (matrix[0]?.[i] === 0) matrix.forEach((row) => (row[i] = 0));
}
```

다만 지극히 제 주관적으로, 위 코드는 전체 라인 수는 줄어들지만 조건이나 처리가 다소 지저분하다고 느껴져서 두 개의 순회로 분리하였습니다!

Reply:
아 네 희찬님, 제가 생각했던 코드는 요런 방식이었습니다! 
이 방법이 가독성이 크게 떨어지지는 않는 것 같아서 소소하게 코멘트 남기려 했습니다!

```suggestion
  for (let i = 1; i < r; i++) {
    for (let j = 1; j < c; j++) {
      if (matrix[i][0] === 0 || matrix[0][j] === 0) {
        matrix[i][j] = 0;
      }
    }
  }
```